File: src/main/java/com/paragon/client/ui/panel/element/setting/StringElement.java
package com.paragon.client.ui.panel.element.setting;

import com.paragon.api.setting.Bind;
import com.paragon.api.setting.Setting;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.client.ui.panel.Click;
import com.paragon.client.ui.panel.element.Element;
import com.paragon.client.ui.panel.element.module.ModuleElement;
import net.minecraft.util.ChatAllowedCharacters;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.text.TextFormatting;
import org.lwjgl.input.Keyboard;

import java.awt.*;

import static org.lwjgl.opengl.GL11.*;

public class StringElement extends Element {

    private final Setting<String> setting;

    private boolean focused;

    private float hover;
    private float listeningFactor;

    public StringElement(int layer, Setting<String> setting, ModuleElement moduleElement, float x, float y, float width, float height) {
        super(layer, x, y, width, height);

        setParent(moduleElement.getParent());
        this.setting = setting;

        setting.getSubsettings().forEach(subsetting -> {
            if (subsetting.getValue() instanceof Boolean) {
                getSubElements().add(new BooleanElement(layer + 1, (Setting<Boolean>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Enum<?>) {
                getSubElements().add(new EnumElement(layer + 1, (Setting<Enum<?>>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Number) {
                getSubElements().add(new SliderElement(layer + 1, (Setting<Number>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Bind) {
                getSubElements().add(new BindElement(layer + 1, (Setting<Bind>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Color) {
                getSubElements().add(new ColourElement(layer + 1, (Setting<Color>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof String) {
                getSubElements().add(new StringElement(layer + 1, (Setting<String>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }
        });
    }

    @Override
    public void render(int mouseX, int mouseY, int dWheel) {
        if (setting.isVisible()) {
            hover = MathHelper.clamp(hover + (isHovered(mouseX, mouseY) ? 0.02f : -0.02f), 0, 1);
            listeningFactor = MathHelper.clamp(listeningFactor + (focused ? 0.02f : -0.02f), 0, 1);

            RenderUtil.drawRect(getX(), getY(), getWidth(), getHeight(), new Color(40, 40, 45).getRGB());
            RenderUtil.drawRect(getX() + getLayer(), getY(), getWidth() - getLayer() * 2, getHeight(), new Color((int) (40 + (30 * hover)), (int) (40 + (30 * hover)), (int) (45 + (30 * hover))).getRGB());
            RenderUtil.drawRect(getX() + getLayer(), getY(), 1, getHeight() * listeningFactor, Color.HSBtoRGB(getParent().getLeftHue() / 360, 1, 0.5f + (0.25f * hover)));

            renderText(setting.getName(), getX() + (getLayer() * 2) + 5, getY() + getHeight() / 2 - 3.5f, 0xFFFFFFFF);

            glPushMatrix();
            glScalef(0.8f, 0.8f, 0.8f);

            {
                float scaleFactor = 1 / 0.8f;

                float side = (getX() + getWidth() - (getStringWidth(getSetting().getValue() + (focused ? "_" : "")) * 0.8f) - 5) * scaleFactor;
                renderText(formatCode(TextFormatting.GRAY) + " " + getSetting().getValue() + (focused ? "_" : ""), side, (getY() + 5f) * scaleFactor, -1);
            }

            glPopMatrix();

            super.render(mouseX, mouseY, dWheel);
        }

        else {
            focused = false;
        }
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, Click click) {
        if (setting.isVisible()) {
            if (isHovered(mouseX, mouseY) && getParent().isElementVisible(this)) {
                if (click.equals(Click.LEFT)) {
                    focused = !focused;
                }
            }

            super.mouseClicked(mouseX, mouseY, click);
        }
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, Click click) {
        if (setting.isVisible()) {
            super.mouseReleased(mouseX, mouseY, click);
        }
    }

    @Override
    public void keyTyped(int keyCode, char keyChar) {
        if (setting.isVisible()) {
            if (focused) {
                if (keyCode == Keyboard.KEY_BACK) {
                    if (getSetting().getValue().length() > 0){
                        getSetting().setValue(getSetting().getValue().substring(0, getSetting().getValue().length() - 1));
                    }
                }

                else if (keyCode == Keyboard.KEY_RETURN) {
                    focused = false;
                }

                else if (ChatAllowedCharacters.isAllowedCharacter(keyChar)) {
                    getSetting().setValue(getSetting().getValue() + keyChar);
                }
            }

            super.keyTyped(keyCode, keyChar);
        }
    }

    @Override
    public float getHeight() {
        return getSetting().isVisible() ? super.getHeight() : 0;
    }

    @Override
    public float getSubElementsHeight() {
        return getSetting().isVisible() ? super.getSubElementsHeight() : 0;
    }

    @Override
    public float getTotalHeight() {
        return getSetting().isVisible() ? super.getTotalHeight() : 0;
    }

    public Setting<String> getSetting() {
        return setting;
    }

    public float getHover() {
        return hover;
    }
}

File: src/main/kotlin/com/paragon/impl/ui/configuration/panel/impl/setting/StringElement.kt
package com.paragon.impl.ui.configuration.panel.impl.setting

import com.paragon.Paragon
import com.paragon.impl.event.client.SettingUpdateEvent
import com.paragon.impl.module.client.ClickGUI
import com.paragon.impl.setting.Setting
import com.paragon.impl.ui.configuration.panel.impl.ModuleElement
import com.paragon.impl.ui.configuration.panel.impl.SettingElement
import com.paragon.impl.ui.util.Click
import com.paragon.util.render.RenderUtil
import com.paragon.util.render.font.FontUtil
import me.surge.animation.Animation
import net.minecraft.util.ChatAllowedCharacters
import org.lwjgl.input.Keyboard
import org.lwjgl.opengl.GL11.*
import java.awt.Color

class StringElement(parent: ModuleElement, setting: Setting<String>, x: Float, y: Float, width: Float, height: Float) : SettingElement<String>(parent, setting, x, y, width, height) {

    val listening = Animation(ClickGUI.animationSpeed::value, false, ClickGUI.easing::value)

    override fun draw(mouseX: Float, mouseY: Float, mouseDelta: Int) {
        super.draw(mouseX, mouseY, mouseDelta)

        RenderUtil.drawRect(x, y, width, height, hover.getColour())

        RenderUtil.scaleTo(x + 3, y + 5.5f, 0f, 0.7, 0.7, 0.7) {
            FontUtil.drawStringWithShadow(setting.name, x + 3, y + 5.5f, Color.WHITE)
        }

        RenderUtil.drawRoundedRect(
            x + getRenderableWidth() - FontUtil.getStringWidth(setting.value) * 0.7f - 8,
            y + 2f,
            FontUtil.getStringWidth(setting.value) * 0.7f + 6,
            height - 4f,
            5f,
            Color(50 + (10 * listening.getAnimationFactor()).toInt(), 50 + (10 * listening.getAnimationFactor()).toInt(), 50 + (10 * listening.getAnimationFactor()).toInt())
        )

        run {
            glScalef(0.7f, 0.7f, 0.7f)

            val scaleFactor = 1 / 0.7f

            val valueX: Float = (x + getRenderableWidth() - FontUtil.getStringWidth(setting.value) * 0.7f - 5) * scaleFactor

            FontUtil.drawStringWithShadow(setting.value, valueX, (y + 5f) * scaleFactor, Color.GRAY)

            glScalef(scaleFactor, scaleFactor, scaleFactor)
        }

        drawSettings(mouseX, mouseY, mouseDelta)
    }

    override fun mouseClicked(mouseX: Float, mouseY: Float, click: Click) {
        super.mouseClicked(mouseX, mouseY, click)

        if (hover.state) {
            if (click == Click.LEFT) {
                listening.state = !listening.state
            } else if (click == Click.RIGHT) {
                expanded.state = !expanded.state
            }

            return
        }
    }

    override fun keyTyped(character: Char, keyCode: Int) {
        super.keyTyped(character, keyCode)

        if (listening.state) {
            if (keyCode == Keyboard.KEY_BACK) {
                if (setting.value.isNotEmpty()) {
                    setting.setValue(setting.value.substring(0, setting.value.length - 1))
                    Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
                }
            }
            else if (keyCode == Keyboard.KEY_RETURN) {
                listening.state = false
            }
            else if (ChatAllowedCharacters.isAllowedCharacter(character)) {
                setting.setValue(setting.value + character)
                Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
            }
        }
    }

}